pretyping/recordops.mli

```diff
@@ -82,20 +88,29 @@ type obj_typ = {
   o_NPARAMS : int;
   o_TCOMPS : constr list } (** ordered *)
 
-(** Return the form of the component of a canonical structure *)
-val cs_pattern_of_constr : Environ.env -> constr -> cs_pattern * int option * constr list
-
-val pr_cs_pattern : cs_pattern -> Pp.t
 
 type cs = GlobRef.t * inductive
 
-val lookup_canonical_conversion : Environ.env -> (GlobRef.t * cs_pattern) -> constr * obj_typ
-val register_canonical_structure : warn:bool -> Environ.env -> Evd.evar_map ->
-  cs -> unit
+val register_canonical_structure :
+  warn:bool -> Environ.env -> Evd.evar_map -> cs -> unit
+
 val subst_canonical_structure : Mod_subst.substitution -> cs -> cs
+
+(** [canonical_projections ()] returns the list of tuples ((p,v),c)
+   Note: p c = v *)
+val canonical_projections : unit -> ((GlobRef.t * cs_pattern) * obj_typ) list
```

Comment:
You should define some parameterized record type for this function and the one below.